chainstate/tx-verifier/src/transaction_verifier/reward_distribution.rs

```diff
@@ -0,0 +1,575 @@
+// Copyright (c) 2023 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::BTreeMap;
+
+use crate::{error::ConnectTransactionError, TransactionSource};
+
+use common::{
+    chain::{Block, DelegationId, PoolId},
+    primitives::{Amount, Id},
+    Uint256,
+};
+use pos_accounting::{
+    AccountingBlockRewardUndo, PoSAccountingOperations, PoSAccountingUndo, PoSAccountingView,
+    PoolData,
+};
+
+use super::accounting_delta_adapter::PoSAccountingDeltaAdapter;
+
+/// Distribute reward among the pool's owner and delegations
+pub fn distribute_pos_reward<P: PoSAccountingView>(
+    accounting_adapter: &mut PoSAccountingDeltaAdapter<P>,
+    block_id: Id<Block>,
+    pool_id: PoolId,
+    total_reward: Amount,
+) -> Result<AccountingBlockRewardUndo, ConnectTransactionError> {
+    let pool_data = accounting_adapter
+        .accounting_delta()
+        .get_pool_data(pool_id)?
+        .ok_or(ConnectTransactionError::PoolDataNotFound(pool_id))?;
+
+    let total_staker_reward = calculate_pool_owner_reward(total_reward, &pool_data).ok_or(
+        ConnectTransactionError::PoolOwnerRewardCalculationFailed(block_id, pool_id),
+    )?;
+
+    let increase_pool_balance_undo = accounting_adapter
+        .operations(TransactionSource::Chain(block_id))
+        .increase_pool_owner_balance(pool_id, total_staker_reward)?;
+
+    let total_delegations_reward = (total_reward - total_staker_reward).ok_or(
+        ConnectTransactionError::PoolOwnerRewardCannotExceedTotalReward(
+            block_id,
+            pool_id,
+            total_staker_reward,
+            total_reward,
+        ),
+    )?;
+
+    let delegation_undos = if total_delegations_reward != Amount::ZERO {
+        distribute_delegations_pos_reward(
+            accounting_adapter,
+            block_id,
+            pool_id,
+            total_delegations_reward,
+        )?
+    } else {
+        None
+    };
+
+    let undos = delegation_undos
+        .unwrap_or_default()
+        .into_iter()
+        .chain(vec![increase_pool_balance_undo].into_iter())
+        .collect();
+
+    Ok(AccountingBlockRewardUndo::new(undos))
+}
+
+fn calculate_pool_owner_reward(total_reward: Amount, pool_data: &PoolData) -> Option<Amount> {
+    (total_reward - pool_data.cost_per_block())
```

Comment:
Given this, we should probably limit the cost_per_block in pool creation.

Reply:
Please create an issue explaining the problem. We can tackle this later.

Reply:
done #868